site/blog/2024-05-19-pathfiinding-part1/pathfindingpart1.md

```diff
@@ -0,0 +1,157 @@
+
+Node B is closer to Source than node D, so we visit it next.
+
+![alt text](./img/image-6.png)
+
+Unvisited neighbors of B are E and F. E is closes to A, so we visit it next.
+
+D is E's unvisited neighbor, but its distance via E is longer than what's already in the result index, so we do not add this data up.
+
+D is the only unvisited neighbor, and we hit a dead end on this branch, so D gets visited, but with no updates to the results table
+
+![alt text](./img/image-7.png)
+
+So since we are looping through all unvisited Nodes, F is the final unvisited node, and its a neighbor of B. We can now visit F through
+B, and we do not have any results table updates with this visit, as F has no unvisited neighbors.
+
+## What do we do with this data?
+
+My library module for using this algorithm includes a method that runs the analysis, then uses the results table to get the shortest
+path.
+
+It takes in the starting node, and ending (destination) node and returns the list of nodes needed to traverse the path.
+
+```ts
+  shortestPath(startnode: Node, endnode: Node): Node[] {
+    let dAnalysis = this.dijkstra(startnode);
+
+    //iterate through dAnalysis to plot shortest path to endnode
+    let path: Node[] = [];
+    let current: Node | null | undefined = endnode;
+    while (current != null) {
+      path.push(current);
+      current = dAnalysis.find(node => node.node == current)?.previous;
+      if (current == null) {
+        break;
+      }
+    }
+    path.reverse();
+    return path;
+  }
+```
+
+So for Example, if I said starting node is A, and endingnode is D, then the returned array would look like.
+
+```ts
+//[Node C, Node D]
+```
+
+If you need the starting node in the path, you can unshift it in.
+
+```ts
+path.unshift(startnode);
+//[Node A, Node C, Node D]
+```
+
+## The test
+
+![Demo Test](./img/image-8.png)
+
+[Link to Demo](https://excaliburjs.com/sample-pathfinding/)
+
+[Link to Github Project](https://github.com/excaliburjs/sample-pathfinding)
+
+The demo is a simple example of using a Excalibur Tilemap and the pathfinding plugin. When the player clicks a tile that does NOT have
+a tree on it, the pathfinding algorithm selected is used to calculate the path. Displayed in the demo is the amount of tiles to
+traverse, and the overall duration of the process required to make the calculation.
+
+Also included, are the ability to add diagonal traversals in the graph. Which simply modifies the graph created with extra edges added,
+please note, diagonal traversal is slightly more expensive than straight up/down, left/right traversal.
+
+## Conclusion
+
+In this article, we reviewed a brief history of Dijkstra's Algorithm, then we created and example graph network and stepped through it
+using the algorithm, and then was able to use it to determine the shortest path of nodes.
+
+This algorithm I have found is more expensive than A\*, but is a nice tool to use when you don't understand the shape and size of the
```

Comment:
It might be nice to have this benefit at the beginning of the article as well

Reply:
this is the only one that i waffled on regarding making a change...